core/src/Telemetry/TelemetryHelper.cs

```diff
@@ -0,0 +1,79 @@
+﻿//----------------------------------------------------------------------
+//
+// Copyright (c) Microsoft Corporation.
+// All rights reserved.
+//
+// This code is licensed under the MIT License.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files(the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions :
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+//------------------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Microsoft.Identity.Core.Telemetry
+{
+    internal class TelemetryHelper : IDisposable
+    {
+        private readonly string _requestId;
+        private readonly EventBase _eventToEnd;
+        private readonly ITelemetry _telemetry;
+        private readonly bool _shouldFlush;
+
+        public TelemetryHelper(ITelemetry telemetry, string requestId, EventBase eventToStart, EventBase eventToEnd, bool shouldFlush)
+        {
+            _telemetry = telemetry;
+            _requestId = requestId;
+            _eventToEnd = eventToEnd;
+            _shouldFlush = shouldFlush;
+            _telemetry.StartEvent(_requestId, eventToStart);
+        }
+
+        #region IDisposable Support
+        private bool _disposedValue = false; // To detect redundant calls
+
+        protected virtual void Dispose(bool disposing)
+        {
+            if (!_disposedValue)
+            {
+                if (disposing)
+                {
+                    _telemetry.StopEvent(_requestId, _eventToEnd);
+                    if (_shouldFlush)
+                    {
+                        _telemetry.Flush(_requestId);
+                    }
+                }
+
+                _disposedValue = true;
+            }
+        }
+
+        // This code added to correctly implement the disposable pattern.
+        public void Dispose()
+        {
+            Dispose(true);

```

Comment:
You should call GC.SupressFinalize(this) here - see details https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/dispose-pattern